src/Elastic.Apm/Metrics/MetricsCollector.cs

```diff
@@ -108,13 +130,18 @@ internal void CollectAllMetricsImpl()
 					if (samplesFromCurrentProvider != null)
 					{
 						var sampleArray = samplesFromCurrentProvider as MetricSample[] ?? samplesFromCurrentProvider.ToArray();
-						if (sampleArray.Any())
+						if (sampleArray.Length != 0 && sampleArray.All(x => !double.IsNaN(x.KeyValue.Value) && !double.IsInfinity(x.KeyValue.Value)))
 						{
 							_logger.Trace()?.Log("Collected {MetricsProviderName} - adding it to MetricSet", metricsProvider.DbgName);
-							samplesFromAllProviders.AddRange(sampleArray);
+							samples.AddRange(sampleArray);
+							metricsProvider.ConsecutiveNumberOfFailedReads = 0;
+						}
+						else
```

Comment:
@SergeyKleyman, @gregkalapos, from comment to `ConsecutiveNumberOfFailedReads` property
https://github.com/elastic/apm-agent-dotnet/blob/c25254bd47b881abb0a91f38e882406f4afa6c45/src/Elastic.Apm/Metrics/IMetricsProvider.cs#L14-L18
an empty list is interpreted as fail read, however, it wasn't so
https://github.com/elastic/apm-agent-dotnet/blob/c25254bd47b881abb0a91f38e882406f4afa6c45/src/Elastic.Apm/Metrics/MetricsCollector.cs#L111-L117

What is the truth?

Reply:
I think the comment should be the truth, which I guess means we should change ``if (samplesFromCurrentProvider != null)` to `if (samplesFromCurrentProvider != null && samplesFromAllProviders.Any())`  and remove this `if (sampleArray.Any()) `.

Maybe 1 thing I can add here is that when we read CPU usage through perf counters then the 1. call does not return a value because it's based on diffing - I'm not sure, maybe this caused some issues.

PS: this was just a quick response, take it with a grain of salt.

Reply:
> PS: this was just a quick response, take it with a grain of salt.

Good, this is what I did, however in a little bit different manner.

Reply:
> So, I believe it isn't an issue due to we need to return NaN values 5 times before we disable collector. 

Right, I think this is fine, the "1 call does not return value"-problem is not relevant here from what I understand.